core/src/main/java/org/springframework/security/oauth2/server/authorization/jose/NimbusJwtEncoder.java

```diff
@@ -0,0 +1,232 @@
+		final JWTClaimsSet jwtClaimsSet = createJwtClaimsSet(claimsSet);
+
+		// Nimbus JWT representation
+		SignedJWT signedJwt = new SignedJWT(jwsHeader, jwtClaimsSet);
+		try {
+			// sign the Nimbus JWT
+			signedJwt.sign(jwsSigner);
+		} catch (JOSEException ex) {
+			throw new JwtException(String.format(ENCODING_ERROR_MESSAGE_TEMPLATE, ex.getMessage()), ex);
+		}
+
+		// Build JWT with compact serialized value, headers and claims set
+		return Jwt.withTokenValue(signedJwt.serialize())
+				.headers(headers -> headers.putAll(joseHeaders.getHeaders()))
+				.claims(claims -> claims.putAll(claimsSet.getClaims()))
+				.build();
+	}
+
+	public void setJoseHeadersCustomizer(Consumer<JoseHeaders.Builder> joseHeadersCustomizer) {
+		this.joseHeadersCustomizer = joseHeadersCustomizer;
+	}
+
+	public void setJwtClaimsSetCustomizer(Consumer<JwtClaimsSet.Builder> jwtClaimsSetCustomizer) {
+		this.jwtClaimsSetCustomizer = jwtClaimsSetCustomizer;
+	}
+
+	@SuppressWarnings("StatementWithEmptyBody")
+	private JWSHeader createJwsHeader(JoseHeaders headers) {
+		JWSHeader.Builder jwsHeaderBuilder = new JWSHeader.Builder(
+				JWSAlgorithm.parse(headers.getHeaderAsString(ALG)));
+		// Parse optional and custom header parameters
+		Map<String, Object> headersMap = headers.getHeaders();
+		try {
+			for (final String name: headersMap.keySet()) {
+				if (ALG.equals(name)) {
+					// skip as alg is already set
+				} else if (TYP.equals(name)) {
+					String typValue = headers.getType();
+					if (typValue != null) {
+						jwsHeaderBuilder = jwsHeaderBuilder.type(new JOSEObjectType(typValue));
+					}
+				} else if (CTY.equals(name)) {
+					jwsHeaderBuilder = jwsHeaderBuilder.contentType(headers.getHeaderAsString(CTY));
+				} else if (CRIT.equals(name)) {
+					List<String> critValues = headers.getHeaderAsStringList(CRIT);
+					if (critValues != null) {
+						jwsHeaderBuilder = jwsHeaderBuilder.criticalParams(new HashSet<>(critValues));
+					}
+				} else if (JKU.equals(name)) {
+					jwsHeaderBuilder = jwsHeaderBuilder.jwkURL(headers.getHeaderAsURI(JKU));
+				} else if (JWK.equals(name)) {
+					String jwkAsString = headers.getHeaderAsString(JWK);
+					if (jwkAsString != null) {
+						jwsHeaderBuilder = jwsHeaderBuilder.jwk(com.nimbusds.jose.jwk.JWK.parse(jwkAsString));
+					}
+				} else if (X5U.equals(name)) {
+					jwsHeaderBuilder = jwsHeaderBuilder.x509CertURL(headers.getHeaderAsURI(X5U));
+				} else if (X5T.equals(name)) {
+					jwsHeaderBuilder = jwsHeaderBuilder.x509CertThumbprint(
+							Base64URL.from(headers.getHeaderAsString(X5T)));
+				} else if (X5T256.equals(name)) {
+					jwsHeaderBuilder = jwsHeaderBuilder.x509CertSHA256Thumbprint(
+							Base64URL.from(headers.getHeaderAsString(X5T256)));
+				} else if (X5C.equals(name)) {
+					jwsHeaderBuilder = jwsHeaderBuilder.x509CertChain(
+							X509CertChainUtils.toBase64List(JSONArrayUtils.parse(headers.getHeaderAsString(X5C))));
+				} else if (KID.equals(name)) {
```

Comment:
How would the caller know this? It seems intrinsically tied to which key is used.